Add protocol and authentication option

phibla opened the pull request:
Using MQTT with username, password, and WebSocket is a practical and secure approach for IoT and real-time applications. 

Security:
Using a username and password adds an authentication layer, preventing unauthorized access to the MQTT broker.
It helps restrict connections to only trusted devices and users.

WebSocket Support:
WebSockets allow MQTT to work over HTTP(S), making it ideal for web applications.
It enables direct communication between IoT devices and web browsers without requiring additional protocols.
Firewall and Network Compatibility:


## Related Issue
Closes #435

aaronzi commented:
Hi @phibla, thank you for contributing this feature.
Could you please sign the ECA with the same email you used in your commits.
https://www.eclipse.org/legal/eca/

phibla commented:
Hi @aaronzi,
thank you for the information. I signed the ECA. All checks have passed now

mdanish98 commented on basyx.aasrepository/basyx.aasrepository-feature-mqtt/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/mqtt/MqttAasRepositoryConfiguration.java:
```diff
@@ -44,10 +44,20 @@ public class MqttAasRepositoryConfiguration {
+	                             @Value("${mqtt.password:}") String password) throws MqttException {
+	    IMqttClient mqttClient = new MqttClient(protocol+"://" + hostname + ":" + port, clientId, new MemoryPersistence());
+	    
+	    MqttConnectOptions options = mqttConnectOptions();
```
I am worried about the configuration of MqttConnectOptions with mqttConnectOptions method, basically MqttConnectOptions is a bean so it is not guaranteed that it is only being used in this method for creating the MQTT Client, so basically the auth related configuration should directly be inside mqttConnectOptions() method instead of mqttClient method. Because irrespective of if security is configured or not the mqttConnectOptions method will always provide the configurations without Security options -> inconsistent and wrong.

phibla replied:
Thank you for your input. I used your suggestion :)

mdanish98 commented on basyx.aasrepository/basyx.aasrepository-feature-mqtt/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/mqtt/MqttAasRepositoryConfiguration.java:
```diff
@@ -44,10 +44,20 @@ public class MqttAasRepositoryConfiguration {
+	    IMqttClient mqttClient = new MqttClient(protocol+"://" + hostname + ":" + port, clientId, new MemoryPersistence());
+	    
+	    MqttConnectOptions options = mqttConnectOptions();
+	    if (!username.isEmpty() && !password.isEmpty()) {
```
Suggestion to use isBlank() instead of isEmpty()

mdanish98 commented on basyx.submodelservice/basyx.submodelservice-feature-mqtt/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/feature/mqtt/MqttSubmodelServiceConfiguration.java:
```diff
@@ -49,12 +49,22 @@ public class MqttSubmodelServiceConfiguration {
-		mqttClient.connect(mqttConnectOptions());
-
-		return mqttClient;
+	public IMqttClient mqttClient(@Value("${mqtt.clientId}") String clientId, 
```
Same comment as above for MqttAasRepositoryConfiguration.java

phibla replied:
Thank you for your input. I used your suggestion :)

mdanish98 reviewed (changes requested):
Hi @phibla ,

Thanks a lot again for your implementation.
I have provided few remarks on the code itself. 
Apart from this I have some general remarks as follows:
- Little bit formatting is needed. Could you please format it a bit?
- Also implementation for classes [MqttSubmodelRepositoryConfiguration ](https://github.com/eclipse-basyx/basyx-java-server-sdk/blob/main/basyx.submodelrepository/basyx.submodelrepository-feature-mqtt/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/mqtt/MqttSubmodelRepositoryConfiguration.java)and [MqttConfiguration ](https://github.com/eclipse-basyx/basyx-java-server-sdk/blob/main/basyx.aasservice/basyx.aasservice-feature-mqtt/src/main/java/org/eclipse/digitaltwin/basyx/aasservice/feature/mqtt/MqttConfiguration.java) is missing, it also needs to be addressed accordingly. Please check if I missed some other classes.
- There is no test to check if it works fine or not, could you please add some tests at the respective test packages?

Please let me know if any points are not clear.

phibla commented:
I refactored the code based on @mdanish98 suggestion. Furthermore, I have added some test to show that the authentication with different protocols and user/password combination works.

mdanish98 commented on basyx.aasrepository/basyx.aasrepository-feature-mqtt/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/mqtt/TestMqttV2AASAggregatorObserver.java:
```diff
@@ -195,4 +197,90 @@ private static Server startBroker() throws IOException {
 	}
+		
+	@Test
+    public void TCPConnectionWithoutCredentials() throws Exception {
```
Method name should start with a lowercase.

phibla replied:
Method name starts now with verb and lowercase like all the other tests

mdanish98 commented on basyx.aasrepository/basyx.aasrepository-feature-mqtt/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/mqtt/TestMqttV2AASAggregatorObserver.java:
```diff
@@ -195,4 +197,90 @@ private static Server startBroker() throws IOException {
+    }
+
+	@Test
+    public void TCPConnectionWitCredentials() throws Exception {
```
Method name should start with a lowercase.

phibla replied:
Method name starts now with verb and lowercase like all the other tests

mdanish98 commented on basyx.aasrepository/basyx.aasrepository-feature-mqtt/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/mqtt/TestMqttV2AASAggregatorObserver.java:
```diff
@@ -195,4 +197,90 @@ private static Server startBroker() throws IOException {
+    }
+
+	@Test
+    public void TCPConnectionWitWrongCredentials() throws Exception {
```
Method name should start with a lowercase.

phibla replied:
Method name starts now with verb and lowercase like all the other tests

mdanish98 commented on basyx.aasrepository/basyx.aasrepository-feature-mqtt/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/mqtt/TestMqttV2AASAggregatorObserver.java:
```diff
@@ -195,4 +197,90 @@ private static Server startBroker() throws IOException {
+    }
+
+	@Test
+    public void WSConnectionWithoutCredentials() throws Exception {
```
Method name should start with a lowercase.

phibla replied:
Method name starts now with verb and lowercase like all the other tests

mdanish98 commented on basyx.aasrepository/basyx.aasrepository-feature-mqtt/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/mqtt/TestMqttV2AASAggregatorObserver.java:
```diff
@@ -195,4 +197,90 @@ private static Server startBroker() throws IOException {
+    }
+
+	@Test
+    public void WSConnectionWitCredentials() throws Exception {
```
Method name should start with a lowercase.

mdanish98 commented on basyx.aasservice/basyx.aasservice-feature-mqtt/src/main/java/org/eclipse/digitaltwin/basyx/aasservice/feature/mqtt/MqttConfiguration.java:
```diff
@@ -41,20 +42,31 @@ public class MqttConfiguration {
 	@ConfigurationProperties(prefix = "mqtt")
-	public MqttConnectOptions mqttConnectOptions() {
+	public MqttConnectOptions mqttConnectOptions(@Value("${mqtt.username:}") String username, @Value("${mqtt.password:}") String password) {
 		MqttConnectOptions mqttConceptOptions = new MqttConnectOptions();
```
There is a typo but for sure not introduced by you and it was there, could you also please fix it?
It should be mqttConnectOptions.

phibla replied:
I fixed the typo

mdanish98 commented on basyx.aasservice/basyx.aasservice-feature-mqtt/src/test/java/org/eclipse/digitaltwin/basyx/aasservice/feature/mqtt/TestMqttAasService.java:
```diff
@@ -30,6 +30,8 @@
 import java.util.Arrays;
 import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
```
These imports on line 33 and 34, I am not sure if they are being used anywhere, could you please check and remove these?

phibla replied:
You are right. It can be removed

mdanish98 commented on basyx.aasservice/basyx.aasservice-feature-mqtt/src/test/java/org/eclipse/digitaltwin/basyx/aasservice/feature/mqtt/TestMqttAasService.java:
```diff
@@ -57,7 +59,10 @@
 import org.eclipse.digitaltwin.basyx.http.BaSyxHTTPConfiguration;
 import org.eclipse.digitaltwin.basyx.http.SerializationExtension;
+import org.eclipse.paho.client.mqttv3.IMqttClient;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
```
Same as above.

phibla replied:
IMqttClinet is used but MqttMessage not. I fixed it

mdanish98 commented on basyx.submodelrepository/basyx.submodelrepository-feature-mqtt/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/mqtt/MqttSubmodelRepositoryConfiguration.java:
```diff
@@ -47,20 +47,31 @@ public class MqttSubmodelRepositoryConfiguration {
 	@ConfigurationProperties(prefix = "mqtt")
-	public MqttConnectOptions mqttConnectOptions() {
+	public MqttConnectOptions mqttConnectOptions(@Value("${mqtt.username:}") String username, @Value("${mqtt.password:}") String password) {
 		MqttConnectOptions mqttConceptOptions = new MqttConnectOptions();
```
Same as above.

phibla replied:
I fixed the typo

mdanish98 commented on basyx.submodelservice/basyx.submodelservice-backend-mongodb/N0E3MTA0QkRBQjU3RTE4NA-FileData-jsonFile1.json:
Is there any specific reasons why this file is added?

phibla replied:
No, it was a temp file created during testing.

mdanish98 commented on basyx.submodelservice/basyx.submodelservice-backend-mongodb/N0E3MTA0QkRBQjU3RTE4NA-FileData-jsonFile2.json:
Is there any specific reasons why this file is added?

phibla replied:
No, it was a temp file created during testing.

mdanish98 commented on basyx.submodelservice/basyx.submodelservice-backend-mongodb/N0E3MTA0QkRBQjU3RTE4NA-SubmodelElementCollection.FileData-jsonFile1.json:
Is there any specific reasons why this file is added?

phibla replied:
No, it was a temp file created during testing.

mdanish98 commented on basyx.submodelservice/basyx.submodelservice-backend-mongodb/N0E3MTA0QkRBQjU3RTE4NA-SubmodelElementCollection.FileData-jsonFile2.json:
Is there any specific reasons why this file is added?

phibla replied:
No, it was a temp file created during testing.

mdanish98 commented on basyx.submodelservice/basyx.submodelservice-feature-mqtt/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/feature/mqtt/MqttSubmodelServiceConfiguration.java:
```diff
@@ -50,29 +50,30 @@ public class MqttSubmodelServiceConfiguration {
 	@ConfigurationProperties(prefix = "mqtt")
-	public MqttConnectOptions mqttConnectOptions() {
+	public MqttConnectOptions mqttConnectOptions(@Value("${mqtt.username:}") String username, @Value("${mqtt.password:}") String password) {
 		MqttConnectOptions mqttConceptOptions = new MqttConnectOptions();
```
Same comment as above for typo.

phibla replied:
Fixed the typo

mdanish98 commented on basyx.submodelservice/basyx.submodelservice-feature-mqtt/src/test/resources/config/password_file.conf:
Is there any specific reasons why this file is added?

phibla replied:
Yes, it has. Moquette is used as the MQTT broker for the test. I had to extend this in the configuration so that I can also test websocket and authentication. A password file is required for authentication, see https://github.com/moquette-io/moquette/blob/main/broker/config/moquette.conf

mdanish98 reviewed (changes requested):
Hi @phibla 

Many thanks for addressing the review remarks. 
I have some minor remarks, please address these well.

Also, below are some general remarks:
- Code formatting is still a little bit misaligned, could you please fix this.
- I put some comments on one class, but not put the same comment on each and every classes, please apply your changes to all the classes where the remarks needs to be addressed.

phibla commented:
Hello @mdanish98,

I have addressed all your points:
- I removed unused imports.
- I refactored the code to ensure there are no formatting issues. There was an inconsistent use of tabs.
- All test methods now start with a verb and use lowercase.
- I renamed the variable from mqttConceptOptions to mqttConnectOptions.
